sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala

```diff
@@ -228,53 +238,79 @@ class MempoolMap(
+    // verified for it. However we still need to check:
+    // - for balance because it is possible that both txs in mem pool and txs from reverted blocks were verified with a
+    //    different balance respect the one restored. The only exception are txs from the oldest reverted block but for
+    //    simplicity they are checked the same.
+    //- for tx size because the size is not a consensus rule so other nodes may have allowed bigger txs in their blocks
+    //- for max nonce gap because the stateNonce was reverted and it is lower than before, so some txs now need to be dropped
 
     val fromAddress = account.asInstanceOf[AddressProposition].address()
     val balance = accountStateReaderProvider.getAccountStateReader().getBalance(fromAddress)
 
     val newExecTxs: mutable.TreeMap[BigInteger, ModifierId] = new mutable.TreeMap[BigInteger, ModifierId]()
     val newNonExecTxs: mutable.TreeMap[BigInteger, ModifierId] = new mutable.TreeMap[BigInteger, ModifierId]()
 
+    val stateNonce = txsFromRejectedBlocks.head.getNonce
+    val maxAcceptableNonce = stateNonce.add(MaxNonceGap.subtract(BigInteger.ONE))
+
     // Recreates from scratch the account's nonExecTxs and execTxs maps, starting from txs from rejected blocks.
-    // They are by default directly added to the execTxs map, because the nonce is surely correct. If a tx is found invalid
-    // for balance, it is discarded. All the subsequent txs, if valid, will become not executable and they will be added
+    // They are by default directly added to the execTxs map, because the nonce is surely not too low. If a tx is found invalid
+    // for balance or size, it is discarded. All the subsequent txs, if valid, will become not executable and they will be added
     // to the nonExec map.
+    // In the (unlikely) case some rejected txs have a nonce gap too big, they and all the subsequent txs will be dropped.
     var destMap = newExecTxs
     var haveBecomeNonExecutable = false
+    var maxNonceGapExceeded = false
 
-    txsFromRejectedBlocks
+    txsFromRejectedBlocks.withFilter(_ => !maxNonceGapExceeded)
       .foreach { tx =>
-        if (tx.size() <= MaxTxSize && balance.compareTo(tx.maxCost) >= 0) {
-          all.put(tx.id, tx)
-          destMap.put(tx.getNonce, tx.id)
-        } else {
-          if (!haveBecomeNonExecutable) {
-            destMap = newNonExecTxs
-            haveBecomeNonExecutable = true
+        if (tx.getNonce.compareTo(maxAcceptableNonce) <= 0) {
+          if (tx.size() <= MaxTxSize && balance.compareTo(tx.maxCost) >= 0) {
+            all.put(tx.id, tx)
+            destMap.put(tx.getNonce, tx.id)
+          } else {
+            if (!haveBecomeNonExecutable) {
+              destMap = newNonExecTxs
+              haveBecomeNonExecutable = true
+            }
           }
         }
+        else if (!maxNonceGapExceeded)
+          maxNonceGapExceeded = true
```

Comment:
```suggestion
        else
          maxNonceGapExceeded = true
```
we may rely on `.withFilter(_ => !maxNonceGapExceeded)` so, as soon we reached this case we will never get inside this part.

Reply:
Good catch! Done